### ISSUE TITLE
samba applets are not functional on target without crystal/oscillator

### ISSUE THREAD

**Issue by majbthrd** [atmelcorp/atmel-software-package#55]
The SAM-BA applets, as used by the SAM-BA tool, are non-functional on a SAMA5D2 target without a crystal or oscillator.

Attempting to use the SAM-BA tool on such a target just results in the ubiquitous:

Error: Could not initialize applet (status: undefined)

The only clue was that the current draw of the SAMA5D2 part is less.

I traced it down to applet_chip_init() in ./samba_applets/common/applet_main.c

This function makes a call to pmc_set_main_oscillator_freq() in ./drivers/peripherals/pmc.c.

The intent of that function seems to be to detect and switch to a different clock source, but if the optional extra hardware is not present, all it seems to do is just leave the target immobilized.

My temporary workaround was to comment out the calls in applet_chip_init().

The mask ROM SAM-BA Monitor is already perfectly functional with the clocks it has already selected.  If this were not so, the SAM-BA tool wouldn't be able to talk to the target, much less download an applet.  My question is: why mess with the clock (and risk preventing customers from downloading code)?  It seems like the SAM-BA applets are perfectly serviceable with the clock left as-is.

Thanks.

**Comment by CyrillePitchen** [comment]
Hi majbthrd,

The pmc_set_main_oscillator_freq() function has been fixed by commits:
- 42fdeba87b0c ("drivers/peripherals/pmc: fix pmc_set_main_oscillator_freq()")
- 5185c79d9052 ("drivers/peripherals/pmc: fix pmc_select_external_osc()")

Those fixes are available in sam-ba 3.2.2 release.

About the reason for changing the clock tree, I tend to agree with you: this should not be needed. I've looked at the git logs to figure out when and why this external osc frequency detection was added: it seems to be there a quite a long time but I didn't find the reason. Maybe it was required by the softpack architecture at some point. For now and until further tests are performed, I would keep the call to pmc_set_main_oscillator_freq() to avoid unexpected side effects and regressions.

Best regards,

Cyrille
